File: src/services/renderers/oscilloscope.ts
import { RenderContext } from './types';

export function renderOscilloscopeVisualizer(ctx: RenderContext) {
  const { ctx: c, width, height, config, timeData } = ctx;
  const centerY = height * 0.52;
  const theme = config.theme;
  const len = timeData.length;
  const sliceWidth = width / (len - 1);

  const passes = [
    { alpha: 0.2, blur: 25, width: 8, color: theme.glowColor },
    { alpha: 0.6, blur: 15, width: 4, color: theme.secondaryColor },
    { alpha: 1.0, blur: 6, width: 2, color: theme.primaryColor }
  ];

  for (const pass of passes) {
    c.save();
    c.globalAlpha = pass.alpha;
    c.shadowBlur = pass.blur;
    c.shadowColor = pass.color;
    c.strokeStyle = pass.color;
    c.lineWidth = pass.width;
    c.beginPath();

    let x = 0;
    for (let i = 0; i < len; i++) {
      const v = (timeData[i] / 128.0) - 1.0;
      const y = centerY + v * (height * 0.3) * config.reactivity.sensitivity;
      if (i === 0) c.moveTo(x, y);
      else c.lineTo(x, y);
      x += sliceWidth;
    }
    c.stroke();
    c.restore();
  }
}

File: src-tauri/src/renderers/styles/oscilloscope.rs
//! Oscilloscope style renderer (`oscilloscope`).

use crate::gpu2d::{Color, Fill, GpuCanvas};
use crate::renderers::{
  theme_glow, theme_primary, theme_secondary, RenderContext,
};

pub fn render(c: &mut GpuCanvas, ctx: &mut RenderContext) {
  let theme = &ctx.config.theme;
  let center_y = ctx.height * 0.52;
  let len = ctx.time_data.len();
  if len < 2 {
    return;
  }
  let slice_width = ctx.width / (len as f32 - 1.0);
  let sensitivity = ctx.config.reactivity.sensitivity;

  let passes: [(f32, f32, f32, Color); 3] = [
    (0.2, 25.0, 8.0, theme_glow(theme)),
    (0.6, 15.0, 4.0, theme_secondary(theme)),
    (1.0, 6.0, 2.0, theme_primary(theme)),
  ];

  for (alpha, blur, width, color) in passes {
    c.save();
    c.set_global_alpha(alpha);
    c.set_shadow(color, blur);
    c.set_stroke(Fill::Solid(color));
    c.set_line_width(width);
    let mut pts: Vec<(f32, f32)> = Vec::with_capacity(len);
    for i in 0..len {
      let v = ctx.time_data[i] as f32 / 128.0 - 1.0;
      let y = center_y + v * (ctx.height * 0.3) * sensitivity;
      pts.push((i as f32 * slice_width, y));
    }
    c.stroke_polyline(&pts);
    c.restore();
  }
}
